source/mir/ndslice/sorting.d

```diff
@@ -679,6 +679,214 @@ version(mir_test_topN)
+`!less(e, slice[k])` (i.e. `slice[k]` is less than or equal to each element to 
+its right according to predicate `less`)))
+
+If `slice` contains equivalent elements, multiple permutations of `slice` may
+satisfy these constraints. In such cases, `pivotPartition` attempts to 
+distribute equivalent elements fairly to the left and right of `k` such that `k`
+stays close to `slice.length / 2`.
+ 
+Params:
+    less = The predicate used for comparison
+    slice = The slice being partitioned
+    pivot = The index of the pivot for partitioning, must be less than 
+    `slice.length` or `0` if `slice.length` is `0`
+Returns:
+    The new position of the pivot
+    
+See_Also:
+    $(HTTP jgrcs.info/index.php/jgrcs/article/view/142, Engineering of a Quicksort
+    Partitioning Algorithm), D. Abhyankar, Journal of Global Research in Computer
+    Science, February 2011. $(HTTPS youtube.com/watch?v=AxnotgLql0k, ACCU 2016
+    Keynote), Andrei Alexandrescu.
++/
+@trusted
+template pivotPartition(alias less = "a < b")
+{
+    import mir.functional: naryFun;
+
+    static if (__traits(isSame, naryFun!less, less))
+    {
+        auto pivotPartition(Iterator, SliceKind kind)
+                (Slice!(Iterator, 1, kind) slice, size_t pivot)
+        {
+            assert(pivot < slice.length || slice.length == 0 && pivot == 0, "pivotPartition: pivot must be less than the length of slice or slice must be empty and pivot zero");
+
+            if (slice.length <= 1) return 0;
+
+            import mir.utility: swapStars;
+
+            // Pivot at the front
+            auto frontI = slice._iterator;
+            auto pivotI = frontI + pivot;
+            swapStars(pivotI, frontI);
+
+            // Fork implementation depending on nothrow copy, assignment, and
+            // comparison. If all of these are nothrow, use the specialized
+            // implementation discussed at 
+            // https://youtube.com/watch?v=AxnotgLql0k.
+            static if (is(typeof(
+                    () nothrow { auto x = slice._iterator; x = slice._iterator; return less(*x, *x); }
+                )))
+            {
+                // Plant the pivot in the end as well as a sentinel
+                auto loI = frontI;
+                auto hiI = frontI + slice.length - 1;
+                auto save = *hiI;
+                *hiI = *frontI; // Vacancy is in r[$ - 1] now
+
+                // Start process
+                for (;;)
+                {
+                    // Loop invariant
+                    version(mir_test_topN)
+                    {
+                        // this used to import std.algorithm.all, but we want to
+                        // save imports when unittests are enabled if possible.
+                        foreach (x; 0 .. (loI - frontI))
+                            assert(!less(*frontI, *(frontI + x)), "pivotPartition: *frontI must not be less than *(frontI + x)");
```

Comment:
`*(frontI + x)` -> `frontI[x]` works much faster for refcounted stuff.

Reply:
Didn't realize `iterator[x]` worked.